src/PgRvn.Server/MessageReader.cs

```diff
@@ -0,0 +1,396 @@
+
+            return (new Execute
+            {
+                PortalName = portalName,
+                MaxRows = maxRowsToReturn
+            }, len);
+        }
+
+        private async Task<(Query, int)> ReadQuery(PipeReader reader, CancellationToken token)
+        {
+            int len = 0;
+
+            var (queryString, queryStringLength) = await ReadNullTerminatedString(reader, token);
+            len += queryStringLength;
+
+            return (new Query
+            {
+                QueryString = queryString
+            }, len);
+        }
+
+        private async Task<(string String, int LengthInBytes)> ReadNullTerminatedString(PipeReader reader, CancellationToken token)
+        {
+            ReadResult read;
+            SequencePosition? end;
+
+            while (true)
+            {
+                read = await reader.ReadAsync(token);
+                end = read.Buffer.PositionOf((byte)0);
+                if (end != null)
+                    break;
+
+                reader.AdvanceTo(read.Buffer.Start, read.Buffer.End);
+            }
+
+            var match = read.Buffer.Slice(0, end.Value);
+            var result = Encoding.UTF8.GetString(match);
+            reader.AdvanceTo(read.Buffer.GetPosition(1, end.Value));
+
+            return (result, (int)match.Length + 1);
+        }
+
+        private async Task<ReadResult> ReadMinimumOf(PipeReader reader, int minSizeRequired, CancellationToken token)
+        {
+            var read = await reader.ReadAsync(token);
+
+            while (read.Buffer.Length < minSizeRequired)
+            {
+                reader.AdvanceTo(read.Buffer.Start, read.Buffer.End);
+                read = await reader.ReadAsync(token);
+            }
+
+            return read;
+        }
+
+        private async Task<byte[]> ReadBytesAsync(PipeReader reader, int length, CancellationToken token)
+        {
+            var read = await ReadMinimumOf(reader, length, token);
+            return ReadBytes(read.Buffer, reader, length);
+        }
+
+        private async Task<int> ReadInt32Async(PipeReader reader, CancellationToken token)
+        {
+            var read = await ReadMinimumOf(reader, sizeof(int), token);
+            return ReadInt32(read.Buffer, reader);
+        }
+
+        private async Task<short> ReadInt16Async(PipeReader reader, CancellationToken token)
+        {
+            var read = await ReadMinimumOf(reader, sizeof(int), token);
+            return ReadInt16(read.Buffer, reader);
+        }
+
+        private async Task<byte> ReadByteAsync(PipeReader reader, CancellationToken token)
+        {
+            var read = await ReadMinimumOf(reader, sizeof(byte), token);
+            return ReadByte(read.Buffer, reader);
+        }
+
+        private byte[] ReadBytes(ReadOnlySequence<byte> readBuffer, PipeReader reader, int length)
+        {
+            var sequence = readBuffer.Slice(0, length);
+            var buffer = new byte[length];
```

Comment:
Here we want to do:
* If value is smaller than 1MB, use ArrayPool
* If value is larger, allocate once and throw away afterward.

Note that this means we have to return the values

Reply:
**If value is smaller than 1MB:**
Why ArrayPool? The parameter is saved in the Bind message instance so who will `ArrayPool.Return` it?

**If value is larger:**
Not sure I understand, I thought you wanted to not accept values over 1MB - which I just implemented. So now if the parameter is too big we skip to the end of the message and throw an error. Why would we allocate over 1MB and throw it away?

Reply:
1. Can `MessageReader` be disposable and return all rented buffers in `Dispose()`? 
2. I'm pretty sure that we need to support values greater than 1MB. Oren's idea was to skip the `ArrayPool` in that case and allocate it using `new byte[length];`. This way we avoid pushing very large buffers to the pool.

@ayende can you check if that's correcT?

Reply:
1. It should be.
2. What is the scenario here. Right now, we want to only allow to send queries. I don't think we'll have very big values there.
3. When we add `exec()` support, we'll need that, though.

Reply:
Even if we don't expect big values here I'd rather support it by doing direct allocation than throwing. If that ever happened that would be very unexpected behavior to the user.

Regarding "What is the scenario here" -  @omerlagziel ?